pkg/destroy/alibabacloud/alibabacloud.go

```diff
@@ -0,0 +1,1343 @@
+	if err != nil {
+		return err
+	}
+
+	o.slbClient, err = slb.NewClientWithOptions(o.Region, config, o.Auth)
+	if err != nil {
+		return err
+	}
+
+	o.ossClient, err = oss.New(ossEndpoint, o.AccessKeyID, o.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+
+	o.rmanagerClient, err = resourcemanager.NewClientWithOptions(o.Region, config, o.Auth)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// New returns an Alibaba Cloud destroyer from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.Destroyer, error) {
+	region := metadata.ClusterPlatformMetadata.AlibabaCloud.Region
+	client, err := icalibabacloud.NewClient(region)
+	if err != nil {
+		return nil, err
+	}
+
+	auth := credentials.NewAccessKeyCredential(client.AccessKeyID, client.AccessKeySecret)
+
+	return &ClusterUninstaller{
+		Logger:          logger,
+		Auth:            auth,
+		AccessKeyID:     client.AccessKeyID,
+		AccessKeySecret: client.AccessKeySecret,
+		Region:          region,
+		InfraID:         metadata.InfraID,
+		ClusterDomain:   metadata.AlibabaCloud.ClusterDomain,
+		Tags: []map[string]string{
+			{
+				fmt.Sprintf("kubernetes.io/cluster/%s", metadata.InfraID): "owned",
+			},
+			{
+				"ack.aliyun.com": metadata.InfraID,
+			},
+		},
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (o *ClusterUninstaller) Run() (*types.ClusterQuota, error) {
+	var err error
+
+	err = o.configureClients()
+	if err != nil {
+		return nil, err
+	}
+
+	err = o.findResources()
+	if err != nil {
+		return nil, err
+	}
+
+	err = o.destroyCluster()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to destroy cluster")
+	}
+
+	return nil, nil
+}
+
+func (o *ClusterUninstaller) destroyCluster() error {
+	stagedFuncs := [][]struct {
+		name    string
+		execute func() error
+	}{
+		{
+			{name: "DNS records", execute: o.deleteDNSRecords},
+			{name: "OSS buckets", execute: o.deleteBucket},
+			{name: "RAM roles", execute: o.deleteRAMRoles},
+			{name: "ECS instances", execute: o.deleteEcsInstances},
+		},
+		{
+			{name: "private zones", execute: o.deletePrivateZones},
+			{name: "ECS security groups", execute: o.deleteSecurityGroups},
+			{name: "Nat gateways", execute: o.deleteNatGateways},
+			{name: "SLBs", execute: o.deleteSlbs},
+		},
+		{
+			{name: "EIPs", execute: o.deleteEips},
+		},
+		{
+			{name: "VSwitchs", execute: o.deleteVSwitchs},
+		},
+		{
+			{name: "VPCs", execute: o.deleteVpcs},
+		},
+		{
+			{name: "resource groups", execute: o.deleteResourceGroup},
+		},
+	}
+
+	for _, stage := range stagedFuncs {
+		var wg sync.WaitGroup
+		errCh := make(chan error)
+		wgDone := make(chan bool)
+
+		for _, f := range stage {
+			wg.Add(1)
+			go o.executeStageFunction(f, errCh, &wg)
+		}
+
+		go func() {
+			wg.Wait()
+			close(wgDone)
+		}()
+
+		select {
+		case <-wgDone:
+			// On to the next stage
+			continue
+		case err := <-errCh:
+			return err
+		}
```

Comment:
If I'm reading this code correctly, an error returned from any of the destroy functions will cause the destroyer to stop. So if we get an error when trying to destroy a DNS record in the first stage, none of the other stages will be executed.

The destroyer cannot stop on an error. It must continuously loop trying to destroy all known resources. It stops when it cannot find more resources to delete.

It looks like this destroy code may be modeled after IBM, which seems to have this same issue. I will bring this up with them to change. 

If you want to follow the current async model you have you can look at the openstack code: https://github.com/openshift/installer/blob/master/pkg/destroy/openstack/openstack.go

Destroy does not have to be async, in which case you can follow the model from AWS.

Reply:
Oops. I see now I was misreading this code. errCh is not passed to the stage functions